### ISSUE TITLE
`ThemeData.copyWith` does not override defaults when `useMaterial3` is true

### ISSUE THREAD

**Issue by guidezpl** [flutter/flutter#105578]
Hmm I think that's a bug then. I would expect `copyWith` with `useMaterial3` true to apply these updates, if I haven't provided my own `typography` and `splashFactory`, of course. 

If I understand correctly, we're currently requiring *every created theme out there* (that doesn't use the factory constructor) to add the following?

```.copyWith(
...
typography: Typography.material2021(platform: platform),
splashFactory: platform == TargetPlatform.android && !kIsWeb ? InkSparkle.splashFactory ?  InkRipple.splashFactory,
)
```

It's not documented anywhere and is pretty brittle.

Nevertheless, that's out of scope for this PR, so I've moved this section above the styles and components sections.

_Originally posted by @guidezpl in https://github.com/flutter/flutter/pull/104815#discussion_r884625696_

**Comment by guidezpl** [comment]
Posting reply from @darrenaustin :

> > Hmm I think that's a bug then. I would expect `copyWith` with `useMaterial3` true to apply these updates, if I haven't provided my own `typography` and `splashFactory`, of course.
> 
> Not a bug so much as an easily misunderstood feature. `copyWith` just blindly overrides any given parameters, but does no other processing. Even if we want to special case this for `useMaterial3`, once a ThemeData is constructed, it will have non-null typography and splashFactory settings. We would not be able to tell if you have provided your own or if they were default.
> 
> > If I understand correctly, we're currently requiring _every created theme out there_ (that doesn't use the factory constructor) to add the following?
> > ```
> > ...
> > typography: Typography.material2021(platform: platform),
> > splashFactory: platform == TargetPlatform.android && !kIsWeb ? InkSparkle.splashFactory ?  InkRipple.splashFactory,
> > )
> > ```
> 
> There is no way to construct a ThemeData that doesn't use the factory constructor other than `ThemeData.raw` or `ThemeData.copyWith` (which itself uses `.raw`). Unless I am missing something here.
> 
> If people want to use `copyWith` in conjunction with `useMaterial3` they will get most of the features, just not the ones that have defaults set in the factory constructor (hence the description in the docs).
> 
> > Nevertheless, that's out of scope for this PR, so I've moved this section above the styles and components sections.
> 
> Thanks.

**Comment by exaby73** [comment]
Hello @guidezpl. Could you post a reproducible example? Although your comments on the PR explains the issue, I'm not sure how to exactly reproduce this in the context of an app.

**Comment by guidezpl** [comment]
As an example, `ThemeData.light().copyWith(useMaterial3: true)` would produce a ThemeData with M2 defaults for typography and splash color.

**Comment by exaby73** [comment]
Yes now I see what you mean. This seems like a feature request to change how `copyWith` works. Though, I believe it would break consistency with how `copyWith` works across APIs. As @darrenaustin said, `copyWith` blindly overrides given parameters. 

What is your proposal for the `copyWith` here? Would be great if you can provide a possible solution on how the API would work :)

**Comment by guidezpl** [comment]
So my proposal is: when `useMaterial3` is false, works as it does now. When `useMaterial3` is true, the special defaults for `splashFactory` and `typography` override the existing default (likely m2, if no customizations have been made) values, unless `splashFactory` or `typography` are provided to `copyWith`.

**Comment by exaby73** [comment]
Alright. Thank for the information :)

**Comment by darrenaustin** [comment]
> So my proposal is: when `useMaterial3` is false, works as it does now. When `useMaterial3` is true, the special defaults for `splashFactory` and `typography` override the existing default (likely m2, if no customizations have been made) values, unless `splashFactory` or `typography` are provided to `copyWith`.

I don't think this is a good proposal. First, as mentioned before it is inconsistent with how all of our `.copyWith` constructors work. They are meant to just override the specific parameters that are given to it, with no more logic than that. Making it inconsistent means it is harder for people to understand our APIs, as this would be a special case.

A more technical issue is that after the constructor has finished initializing, the default typography and splash factory are fully initialized to non-null values. In `.copyWith` there is no way to tell if the values for typography and splash factory were set by the user in the constructor or just set to the  defaults. We could check to see if the values are equal to the defaults, but even then the user could have specified the defaults in the constructor themselves and we would be overriding their choice with this proposal (admittedly this is unlikely, but possible). Another wrinkle to this is that the splash factory default logic is a little more complicated as it depends on the platform (i.e. sparkle factory for non-web Android devices, ripple for everything else), so we would need to replicate that in `.copyWith` for this check.

While I am sympathetic to the issue of `useMaterial3` only effecting the defaults for these two properties in the constructor and not `.copyWith` seems counter-intuitive, I think this alternative is problematic.

**Comment by guidezpl** [comment]
> I don't think this is a good proposal. First, as mentioned before it is inconsistent with how all of our .copyWith constructors work. They are meant to just override the specific parameters that are given to it, with no more logic than that. Making it inconsistent means it is harder for people to understand our APIs, as this would be a special case.

I understand your point about `copyWith` but this is a special case. IMO, having special defaults for `ThemeData` as well as having an opt-in flag which only partially opts you in when you use `copyWith` (which is super common) makes the API harder to understand.

The technical issues are fairly surmountable. 

> While I am sympathetic to the issue of useMaterial3 only effecting the defaults for these two properties in the constructor and not .copyWith seems counter-intuitive, I think this alternative is problematic.

If we stay this route and don't want every app which opts-in to Material 3 with a Material 2 constructed theme to have the wrong typography and ink splash, we should consider: 
- clearly documenting how to set the right typography and splashFactory for these themes
- dealing with named constructors, which currently cannot properly opt-in to Material 3
  - adding `useMaterial3` to `light`, `dark`, `fallback` constructors
  - deprecating named constructors

**Comment by exaby73** [comment]
One thing to note about the consistency... Using `ThemeData().copyWith(useMaterial3: true)`, according to the logic of `copyWith` is meant to override specific parameters with no logic, there wouldn't be any change in say overscroll behavior either. Though with just `ThemeData()`, so M2 defaults, we get the glow effect while with `ThemeData().copyWith(useMaterial3: true)`, we get the stretch effect on Android 12. I would then expect, as a user, that changing that parameter wouldn't change the already set defaults. 

To me, and take this with a grain of salt (I am in no way an expert here), `copyWith` is already causing other defaults to change even though it's not doing it directly. The way it looks like, this flag `useMaterial3`, shouldn't even be a parameter on `copyWith` since it's indirectly causing a whole bunch of changes to other defaults which kind of goes against how `copyWith` should function. Again, I'm no expert and I may be wrong here and please correct me if I am.

**Comment by rydmike** [comment]
I would also agree with @exaby73, the `useMaterial3` should perhaps not have been included in `copyWith`, at least not considering how the current `ThemeData` factory is designed.

To get all the right behaviour for opt-in/out of M3, you need to use the ThemeData factory and set it there. If you change it after that, you will get conflicting behaviour in any default (null props) component themes, that check this property for how they should look/behave. If all the behaviour props where in their own sub-themes, that checked the value of `useMaterial3` prop, then it would work, but as some behaviour is defined by the ThemeData factory, the only current way to get the correct and consistent behaviour is to create a new ThemeData object with the factory constructor, using copyWith will not do it.

I think the doc comment on the `useMaterial3` flag is very clear on this point:

> If a [ThemeData] is constructed with [useMaterial3] set to true, then some properties will get special defaults. However, just copying a [ThemeData] with [useMaterial3] set to true will not change any of these properties in the resulting [ThemeData]. These properties are:

even if it might feel confusing at first.

**Comment by exaby73** [comment]
Maybe the approach would be to remove the `useMaterial3` flag altogether and instead have a `ThemeData.material3()` factory instead. Thoughts?

**Comment by rydmike** [comment]
Since the flag is intended to be a temporary one that only exist during migration to M3, and M3 design will eventually become the only option for Material themed apps, it is perhaps not an ideal solution either. I'm just speculating, not a member of the Flutter/Material team, just another Flutter community member with a passion for theming, lol :)

**Comment by bernaferrari** [comment]
This is really small, but older Android devices (80%) doesn't have the sparkle effect and Samsung (probably 50% of Android devices?) doesn't have the sparkle effect even in Android 12. So the effect would be different from what users expect.

**Comment by guidezpl** [comment]
Thanks for the discussion, which was useful in planning forward, will close this as all further work is integrated in the planning for expressive.
